src/main/java/uk/ac/ebi/ampt2d/accession/service/MonotonicAccessionGenerator.java

```diff
@@ -0,0 +1,241 @@
+        if (lastBlock == null) {
+            addBlock(contiguousIdBlockRepository.save(new ContiguousIdBlock(categoryId, instanceId, 0, size)));
+        } else {
+            addBlock(contiguousIdBlockRepository.save(lastBlock.nextBlock(instanceId, size)));
+        }
+    }
+
+    private void addBlock(ContiguousIdBlock block) {
+        activeBlocks.offer(block);
+        availableRanges.add(new MonotonicRange(block.getLastCommitted() + 1, block.getEnd()));
+    }
+
+    public MonotonicRangePriorityQueue getAvailableRanges() {
+        return availableRanges;
+    }
+
+    public synchronized void recoverState(long[] committedElements) throws AccessionIsNotPending {
+        List<MonotonicRange> ranges = MonotonicRange.convertToMonotonicRanges(committedElements);
+        List<MonotonicRange> newAvailableRanges = new ArrayList<>();
+        for (MonotonicRange monotonicRange : this.availableRanges) {
+            newAvailableRanges.addAll(monotonicRange.excludeIntersections(ranges));
+        }
+
+        this.availableRanges.clear();
+        this.availableRanges.addAll(newAvailableRanges);
+        doCommit(committedElements);
+    }
+
+    public synchronized void commit(long... accessions) throws AccessionIsNotPending {
+        assertAccessionIsPending(accessions);
+
+        doCommit(accessions);
+    }
+
+    private void assertAccessionIsPending(long[] accessions) throws AccessionIsNotPending {
+        for(long accession: accessions){
+            if(!generatedIds.contains(accession)){
+                throw new AccessionIsNotPending(accession);
+            }
+        }
+    }
+
+    private synchronized void doCommit(long[] accessions) {
+        addToCommited(accessions);
+
+        ContiguousIdBlock block = activeBlocks.peek();
+        long lastCommitted = block.getLastCommitted();
+        while (committed.peek() != null && lastCommitted + 1 == committed.peek()) {
+            lastCommitted++;
+            committed.poll();
+            if (lastCommitted == block.getEnd()) {
+                activeBlocks.poll();
+                updateBlockLastCommited(block, lastCommitted);
+                block = activeBlocks.peek();
+                lastCommitted = block.getLastCommitted();
+            }
+        }
+
+        if (lastCommitted != block.getLastCommitted()) {
+            updateBlockLastCommited(block, lastCommitted);
+        }
+    }
+
+    private void addToCommited(long[] accessions) {
+        for (long accession : accessions) {
+            committed.offer(accession);
+            generatedIds.remove(accession);
+        }
+    }
+
+    private void updateBlockLastCommited(ContiguousIdBlock block, long lastCommitted) {
+        block.setLastCommitted(lastCommitted);
+        contiguousIdBlockRepository.save(block);
+    }
+
+    public synchronized void release(long... accessions) throws AccessionIsNotPending {
+        assertAccessionIsPending(accessions);
```

Comment:
I'm not sure if this is as easy as for the `commit` method, but would it be possible for this method to use `generatedIds` so we can avoid this check?

Reply:
Same as before